llvm/include/llvm/Support/Compiler.h

```diff
@@ -129,14 +129,72 @@
 #if (!(defined(_WIN32) || defined(__CYGWIN__)) ||                              \
      (defined(__MINGW32__) && defined(__clang__)))
 #define LLVM_LIBRARY_VISIBILITY LLVM_ATTRIBUTE_VISIBILITY_HIDDEN
-#if defined(LLVM_BUILD_LLVM_DYLIB) || defined(LLVM_BUILD_SHARED_LIBS)
-#define LLVM_EXTERNAL_VISIBILITY LLVM_ATTRIBUTE_VISIBILITY_DEFAULT
-#else
-#define LLVM_EXTERNAL_VISIBILITY
-#endif
+#define LLVM_ALWAYS_EXPORT LLVM_ATTRIBUTE_VISIBILITY_DEFAULT
+#elif defined(_WIN32)
+#define LLVM_ALWAYS_EXPORT __declspec(dllexport)
+#define LLVM_LIBRARY_VISIBILITY
 #else
 #define LLVM_LIBRARY_VISIBILITY
-#define LLVM_EXTERNAL_VISIBILITY
+#define LLVM_ALWAYS_EXPORT
+#endif
+
+/// LLVM_ABI is the main export/visibility macro to mark something as explicitly
+/// exported when llvm is built as a shared library with everything else that is 
+/// unannotated will have internal visibility.
+/// 
+/// LLVM_EXPORT_TEMPLATE is used on explicit template instantiations in source
+/// files that were declared extern in a header. This macro is only set as a 
+/// compiler export attribute on windows, on other platforms it does nothing.
+/// 
+/// LLVM_TEMPLATE_ABI is for annotating extern template declarations in headers 
+/// for both functions and classes. On windows its turned in to dllimport for 
+/// library consumers, for other platforms its a default visibility attribute.
+/// 
+/// LLVM_C_ABI is used to annotated functions and data that need to be exported 
+/// for the libllvm-c API. This used both for the llvm-c headers and for the 
+/// functions declared in the different Target's c++ source files that don't include
+/// the header forward declaring them.
+#ifndef LLVM_ABI_GENERATING_ANNOTATIONS
+// Marker to add to classes or functions in public headers that should not have
+// export macros added to them by the clang tool
+#define LLVM_ABI_NOT_EXPORTED
+#if defined(LLVM_BUILD_LLVM_DYLIB) || defined(LLVM_BUILD_SHARED_LIBS)
+#if defined(_WIN32)
+#if defined(LLVM_ABI_EXPORTS)
+#define LLVM_ABI __declspec(dllexport)
+#define LLVM_TEMPLATE_ABI
+#define LLVM_EXPORT_TEMPLATE LLVM_ABI
+#elif defined(LLVM_DLL_IMPORT)
+#define LLVM_ABI __declspec(dllimport)
+#define LLVM_TEMPLATE_ABI __declspec(dllimport)
+#define LLVM_EXPORT_TEMPLATE
+#else
```

Comment:
This mode is needed for a few static libraries like tablegen don't link to the llvm shared library and if it was left exporting symbols there might be duplicate symbols errors from two different llvm static libraries linked in to an executable.

Reply:
Can you add a comment to that effect here after the `#else`?

Reply:
Can we use the following pattern instead?

```c
#if defined(LLVM_BUILD_STATIC)
# define LLVM_ABI
# define LLVM_TEMPLATE_ABI
# define LLVM_EXPORT_TEMPLATE
#else
# if defined(LLVM_EXPORTS)
#   define LLVM_ABI __declspec(dllexport)
#   define LLVM_TEMPLATE_ABI
#   define LLVM_EXPORT_TEMPLATE __declspec(dllexport)
# else
#   define LLVM_ABI __declspec(dllimport)
#   define LLVM_TEMPLATE_ABI __declspec(dllimport)
#   define LLVM_EXPORT_TEMPLATE
# endif
#endif
```

If you are statically linking, you define `LLVM_BUILD_STATIC`, otherwise, if you are building the DLL, you define `LLVM_EXPORTS`.

Reply:
The purpose of doing it that way is to make it so normal users don't have to define macros to get the right settings. But, those dllimport settings are incorrect if you're linking against LLVM statically. It raises the question of, in the future, which will be more common, using LLVM.dll, or static linking? I DLLs are better for distribution, so we should lean into and go with @compnerd's suggested direction.

Reply:
i wasn't sure if @compnerd meant setting LLVM_BUILD_STATIC for specific libraries like tablegen when building llvm as a shared library.

Reply:
There is also the weird edge case of statically linking and LLVM_EXPORT_SYMBOLS_FOR_PLUGINS where component symbols are exported from the tool executables.

Reply:
@rnk exactly right :) Most users wouldn't need to figure out defines, if they want static linking they specify `/DLLVM_BUILD_STATIC` (we can rename that to `/DLLVM_STATIC` or `/DLLVM_STATIC_BUILD` or something along those lines). Otherwise, they get the default behaviour of DLL linkage.

Reply:
@compnerd i had push some changes to reorganize the macros